src/Microsoft.AspNetCore.Mvc.Formatters.Json/MvcJsonOptionsExtensions.cs

```diff
@@ -0,0 +1,69 @@
+﻿// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+using System;
+using Microsoft.AspNetCore.Mvc;
+using Microsoft.AspNetCore.Mvc.Formatters.Json;
+using Newtonsoft.Json.Serialization;
+
+namespace Microsoft.Extensions.DependencyInjection
+{
+    public static class MvcJsonOptionsExtensions
+    {
+        /// <summary>
+        /// Configures the casing behavior of JSON serialization to use camel case for property names, 
+        /// and optionally for dynamic types and dictionary keys.
+        /// </summary>
+        /// <param name="options"><see cref="MvcJsonOptions"/></param>
+        /// <param name="processDictionaryKeys">If true will camel case dictionary keys and properties of dynamic objects.</param>
+        /// <returns><see cref="MvcJsonOptions"/> with camel case settings</returns>
+        public static MvcJsonOptions UseCamelCasing(this MvcJsonOptions options, bool processDictionaryKeys)
+        {
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+            
+            if (options.SerializerSettings.ContractResolver.GetType() != typeof(DefaultContractResolver))
+            {
+                var contractResolverName = options.SerializerSettings.ContractResolver.GetType().Name;
+                throw new InvalidOperationException(
+                    Resources.InvalidContractResolverForJsonCasingConfiguration(contractResolverName));
+            }
+
+            var resolver = (DefaultContractResolver)options.SerializerSettings.ContractResolver;
+            resolver.NamingStrategy = new CamelCaseNamingStrategy
+            {
+                ProcessDictionaryKeys = processDictionaryKeys
+            };
+
+            return options;
+        }
+
+        /// <summary>
+        /// Configures the casing behavior of JSON serialization to use the member's casing for property names, 
+        /// properties of dynamic types, and dictionary keys.
+        /// </summary>
+        /// <param name="options"><see cref="MvcJsonOptions"/></param>
+        /// <returns><see cref="MvcJsonOptions"/> with member casing settings.</returns>
+        public static MvcJsonOptions UseMemberCasing(this MvcJsonOptions options)
+        {
+            if (options == null)
+            {
+                throw new ArgumentNullException(nameof(options));
+            }
+
+            if (options.SerializerSettings.ContractResolver.GetType() != typeof(DefaultContractResolver))
```

Comment:
Why use `GetType` and casting when you can use C#'s "pattern matching" like I showed earlier?

Reply:
By using pattern matching the code would not hit the InvalidOperationException path if contract inherits from DefaultContractResolver. I've two tests that display it: 
- UseCamelCasing_WillThrow_IfContractResolver_IsNot_DefaultContractResolver
- UseMemberCasing_WillThrow_IfContractResolver_IsNot_DefaultContractResolver

Reply:
```csharp
if (options.SerializerSettings.ContractResolver is DefaultContractResolver resolver)
{
    // Do whatever with resolver
}
else
{
    // Throw
}
```

Wouldn't that work?

Reply:
I think @AlexejTimonin is attempting to match the exact type (while ignoring derived types). Pattern matching wouldn't work in that case.

Reply:
Not if ContractResolver inherits from DefaultContractResolver, for example CamelCasePropertyNamesContractResolver would not hit the else branch.

Reply:
`is` is preferred when testing the type of an instance

Reply:
Ah, I see. I just didn't understand why that was important. As @JamesNK said (emphasis mine)

> I wonder if these methods should just throw if the type isn't DefaultContractResolver.
> 
> 99% of people will always have that type, **or a subclass of it**, assigned. And for the 1% who have set something different, having these methods reassign a new instance of DefaultContractResolver back likely wouldn't be what they want.

I understood that to mean "throw unless it is or inherits from `DefaultContractResolver`"

Reply:
Granted, it would be pretty weird to have an instance of `CamelCasePropertyNamesContractResolver` configured and then call `UseMemberCasing`, but I don't think it's necessary to throw?

Reply:
Lets take `CamelCasePropertyNamesContractResolver` as an example. In its default constructor it sets `ProcessDictionaryKeys` and `OverrideSpecifiedNames` to `true`. 
The `UseCamelCase` extension method creates new `CamelCaseNamingStrategy` and sets `ProcessDictionaryKeys` to whatever the user supplies, but `OverrideSpecifiedNames` as false.
Now we have configuration collision if the user does something like this: 

            services
                .AddMvc()
                .AddJsonOptions(options =>
                {
                    options.SerializerSettings.ContractResolver = new CamelCasePropertyNamesContractResolver();
                    options.UseCamelCasing(processDictionaryKeys: true);
                });

Should the framework tell the developer that she may encounter problems, or should the framework apply the configuration in order and rely on the developer to detect eventual problems / bugs?

Reply:
Let's do the simple things and check for `DefaultContractResolver` and all subtypes for it (the `is` check) . We can always change the code if we get reports from users that the behavior is odd. If it's worthwhile, feel free to add a remark that says this modifies the `ContractResolver` so users are aware of what this does.

Reply:
I agree. Use `is`.

Assigning `CamelCasePropertyNamesContractResolver` and then calling `UseMemberCasing` would disable the camel case resolver, and would be an odd thing to do, but it would be immediately obvious if that isn't what you wanted and easily fixed.

Reply:
Thank you all for the great feedback! I will make changes this weekend.